aea/helpers/yaml_utils.py

```diff
@@ -0,0 +1,157 @@
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Helper functions related to YAML loading/dumping."""
+import os
+import re
+from collections import OrderedDict
+from typing import Any, Dict, List, Match, Optional, Sequence, TextIO, cast
+
+import yaml
+from yaml import MappingNode
+
+
+class _AEAYamlLoader(yaml.SafeLoader):
+    """
+    Custom yaml.SafeLoader for the AEA framework.
+
+    It extends the default SafeLoader in two ways:
+    - loads YAML configurations while *remembering the order of the fields*;
+    - resolves the environment variables at loading time.
+
+    This class is for internal usage only; please use
+    the public functions of the module 'yaml_load' and 'yaml_load_all'.
+    """
+
+    envvar_matcher = re.compile(r"\${([^}^{]+)\}")
+    envvar_key = "!envvar"
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the AEAYamlLoader.
+
+        It adds a YAML Loader constructor to use 'OderedDict' to load the files.
+        """
+        super().__init__(*args, **kwargs)
+        _AEAYamlLoader.add_constructor(
+            yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG, self._construct_mapping
+        )
+        _AEAYamlLoader.add_constructor(self.envvar_key, self._envvar_constructor)
+        self._add_implicit_resolver_if_not_present_already()
+
+    def _add_implicit_resolver_if_not_present_already(self) -> None:
+        """Add implicit resolver for environment variables, if not present already."""
+        if self.envvar_key in dict(self.yaml_implicit_resolvers.get(None, [])):
+            _AEAYamlLoader.add_implicit_resolver(
+                self.envvar_key, self.envvar_matcher, None
+            )
+
+    @staticmethod
+    def _construct_mapping(loader: "_AEAYamlLoader", node: MappingNode):
+        """Construct a YAML mapping with OrderedDict."""
+        object_pairs_hook = OrderedDict
+        loader.flatten_mapping(node)
+        return object_pairs_hook(loader.construct_pairs(node))
+
+    @staticmethod
+    def _envvar_constructor(
+        _loader: "_AEAYamlLoader", node: MappingNode
+    ):  # pragma: no cover
+        """Extract the matched value, expand env variable, and replace the match."""
+        node_value = node.value
+        match = _AEAYamlLoader.envvar_matcher.match(node_value)
+        match = cast(Match[str], match)
+        env_var = match.group()[2:-1]
+
+        # check for defaults
+        var_name, default_value = env_var.split(":")
+        var_name = var_name.strip()
+        default_value = default_value.strip()
+        var_value = os.getenv(var_name, default_value)
+        return var_value + node_value[match.end() :]
```

Comment:
beware: the following will be applied to _any_ YAML field. That means also non-configurable ones because currently validation is done _after_ custom YAML loading. This should be reported as issue. It could be easily solved by first loading normally and then loading it with replacement. Or, just doing the variable replacement by hand only for configurable fields...

Reply:
see https://github.com/fetchai/agents-aea/blob/master/aea/configurations/loader.py#L262 and https://github.com/fetchai/agents-aea/blob/master/aea/configurations/loader.py#L327

Reply:
I suspect this is linked to the broader issue of having both descriptive and configurable fields in the same file. If we had one "manifest" file and a "configuration" field we would only apply this approach to the latter file.